k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientPresenter.java

```diff
@@ -766,6 +767,51 @@ public void messageReceived() {
         }
     }
 
+    public void handleResetPartnerKeyResult() {
+        loadPEpStatus();
+    }
+
+    public boolean canResetSenderKeys(
+            List<Address> newToAdresses,
+            List<Address> newCcAdresses,
+            List<Address> newBccAdresses
+    ) {
+        return recipientConditionsForKeyReset(newToAdresses, newCcAdresses, newBccAdresses)
+                && ratingConditionsForSenderKeyReset();
+    }
+
+    private void handleSingleAddressKeyResetAllowance(
+            List<Address> newToAdresses,
+            List<Address> newCcAdresses,
+            List<Address> newBccAdresses
+    ) {
+        if (canResetSenderKeys(newToAdresses, newCcAdresses, newBccAdresses)) {
+            recipientMvpView.showResetPartnerKeyOption();
+        } else {
+            recipientMvpView.hideResetPartnerKeyOption();
+        }
+    }
+
+    private boolean recipientConditionsForKeyReset(
+            List<Address> newToAdresses,
+            List<Address> newCcAdresses,
+            List<Address> newBccAdresses
+    ) {
+        return account != null && account.isPlanckPrivacyProtected()
+                && newToAdresses.size() == ONE_ADDRESS
+                && newBccAdresses.isEmpty()
+                && newCcAdresses.isEmpty()
+                && !Preferences.getPreferences(context).containsAccountByEmail(newToAdresses.get(0).getAddress());
+    }
+
+    private boolean ratingConditionsForSenderKeyReset() {
+        return !PlanckUtils.isRatingUnsecure(privacyState) || (privacyState == Rating.pEpRatingMistrust);
```

Comment:
do we really want to prevent reset for privacy state unsecure? How do we solve errors then?

Reply:
Well according to the definition of the method:
![image](https://github.com/plancksecurity/android-planckForAndroid/assets/151338104/97b8ee4f-881c-482f-8038-e1bc5964d888)
- pEpRatingUnderAttack is a rating that should never key-reset, this is something really bad like we are under attack.
- pEpRatingB0rken is the case of a message that breaks the protocol, this is considered an unsafe message so wwe do not allow any operations on it.
- pEpRatingUndefined is a default rating coming from the core that we should never see, it is an "internal error" rating so to speak. So we should not allow key reset here.
- pEpRatingCannotDecrypt -> We cannot decrypt so there is no key to be reset.
- pEpRatingHaveNoKey -> same as pEpRatingCannotDecrypt
- pEpRatingUnencrypted -> message is not encrypted so no key to be reset. I do not believe we reset attached keys in TOFU messages, that seems so dangerous.
- pEpRatingMistrust is the only unsecure rating in which we actually allow key reset, because this is done by the very user who could have rejected partner trustwords by mistake.

Reply:
ok all clear, I just wonder how will under attack be resolved, but that's fine as an explaination

Reply:
yes, we need to actually try these ratings.